frontend/packages/data-portal/app/components/Viewer/state.ts

```diff
@@ -301,6 +304,123 @@ export function toggleOrMakeDimensionPanel() {
   else updateState(toggleDimensionPanelVisible)
 }
 
+export function isTomogramActivatedFromConfig(
+  tomogramConfig: string | undefined | null,
+) {
+  if (!tomogramConfig) return false
+  const layers = currentNeuroglancerState().layers || []
+  const jsonConfig = JSON.parse(tomogramConfig) as NeuroglancerState
+  const newLayers = jsonConfig.layers || []
+  const tomogramLayer = newLayers.find((l) => l.type === 'image')
+  if (!tomogramLayer) return false
+  return layers.some(
+    (l) =>
+      l.type === 'image' &&
+      getLayerSourceUrl(l) === getLayerSourceUrl(tomogramLayer),
+  )
+}
+
+export function isTomogramActivated(tomogramPath: string | undefined | null) {
+  if (!tomogramPath) return false
+  const layers = currentNeuroglancerState().layers || []
+  return layers.some(
+    (l) => l.type === 'image' && getLayerSourceUrl(l) === tomogramPath,
+  )
+}
+
+function inferVoxelSpacingFromState(state: NeuroglancerState) {
+  const { dimensions } = state
+  if (dimensions === undefined) {
+    throw new Error('Cannot infer voxel spacing without dimensions')
+  }
+  // Get the average of all dims, usually isotropic but just in case
+  const dimensionValues = Object.values(dimensions)
+  const averageUnit =
+    dimensionValues.reduce((a: number, b: DimensionValue) => a + b[0], 0) /
+    dimensionValues.length
+  return averageUnit
+}
+
+export function replaceOnlyTomogramSource(
+  incomingState: NeuroglancerState,
+  newPath: string,
+) {
+  const newState = incomingState
+  const tomogramLayer = newState.layers?.find((l) => l.type === 'image')
+  if (tomogramLayer) {
+    // Replace either the source directly or the url inside the source object
+    if (typeof tomogramLayer.source === 'string') {
+      tomogramLayer.source = newPath
+    } else {
+      tomogramLayer.source.url = newPath
+    }
+  }
+  return newState
+}
+
+export function replaceOnlyTomogram(
```

Comment:
Maybe the difference between the two tomogram replace could be clearer. The onlySource one is much simpler and can only be used if we know the voxel spacings and alignments between the two spaces matches